src/scipp/core/data_group.py

```diff
@@ -0,0 +1,192 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2022 Scipp contributors (https://github.com/scipp)
+# @author Simon Heybrock
+from __future__ import annotations
+
+import functools
+import operator
+from collections.abc import MutableMapping
+from typing import Callable, Iterable
+
+from .cpp_classes import DataArray, Dataset, Variable
+
+
+def _item_dims(item):
+    return getattr(item, 'dims', ())
+
+
+def _summarize(item):
+    if isinstance(item, DataGroup):
+        return f'{type(item).__name__}({len(item)}, {item.sizes})\n'
+    if hasattr(item, 'sizes'):
+        return f'{type(item).__name__}({item.sizes})\n'
+    return str(item)
+
+
+class DataGroup(MutableMapping):
+    """
+    A group of data. Has dims and shape, but no coords.
+    """
+
+    def __init__(self, items=None):
+        self._items = {}
+        if items is not None:
+            for name, item in items.items():
+                self[name] = item
+
+    def __len__(self) -> int:
+        return len(self._items)
+
+    def __iter__(self):
+        yield from self._items
+
+    def __getitem__(self, name):
+        if isinstance(name, str):
+            return self._items[name]
+        if name == ():
+            return DataGroup({key: var[()] for key, var in self.items()})
+        dim, index = name
+        if isinstance(index, int) and self.sizes[dim] is None:
+            raise ValueError(
+                f"Positional indexing dim {dim} not possible as the length is not "
+                "unique.")
+        return DataGroup({
+            key: var[dim, index] if dim in _item_dims(var) else var
+            for key, var in self.items()
+        })
+
+    def __setitem__(self, name, value):
+        self._items[name] = value
+
+    def __delitem__(self, name: str):
+        del self._items[name]
+
+    @property
+    def dims(self):
+        dims = ()
+        for var in self.values():
+            # Preserve insertion order
+            for dim in _item_dims(var):
+                if dim not in dims:
+                    dims = dims + (dim, )
+        return dims
+
+    @property
+    def ndim(self):
+        return len(self.dims)
+
+    @property
+    def shape(self):
+
+        def dim_size(dim):
+            sizes = []
+            for var in self.values():
+                if dim in _item_dims(var):
+                    sizes.append(var.sizes[dim])
+            sizes = set(sizes)
+            if len(sizes) == 1:
+                return next(iter(sizes))
+            return None
+
+        return tuple(dim_size(dim) for dim in self.dims)
+
+    @property
+    def sizes(self):
+        return dict(zip(self.dims, self.shape))
+
+    def _make_html(self):
+        out = ''
+        for name, item in self.items():
+            if isinstance(item, DataGroup):
+                html = item._make_html()
+            elif isinstance(item, (Variable, DataArray, Dataset)):
```

Comment:
Is this intentional? `_summarize` seems to be meant to handle those 3 types as well.

Reply:
All the reprs are just a hack to have something working, I expect this to be completely replaced in a follow-up.